#166272288 Set Up React Project

GerrardE opened the pull request:
#### What does this PR do?
It sets up a boilerplate react project

#### Description of Task to be completed?
- Install dependencies and development dependencies
- Configure webpack
- Configure githint
- Setup folder structure

#### How should this be manually tested?
1. clone this repo
2. run "npm install"
3. run "npm run dev"

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#166272288](https://www.pivotaltracker.com/story/show/166272288)

#### Screenshots (if appropriate)
N/A

#### Questions:
N/A

GerrardE committed:
ch-[166272288]: Set up react project
>
>
> - Configure webpack
> - Configure githint
- Setup folder structure

[#166272288]

GerrardE committed:
ch-[166272288]: Set up react project
>
>
> - Refactor webpack configurations for production and development

[#166272288]

GerrardE committed:
ch-[166272288]: Set up react project
>
>
> - Refactor webpack configurations for production and development using DRY principle

[#166272288]

fidelisojeah commented on webpack.base.js:
```diff
@@ -0,0 +1,52 @@
+  module: {
+    rules: [
+      {
+        test: /\.jsx?$/,
```
what about just `.js` files?

fidelisojeah replied:
also modify this to complain on `eslint` violations.

fidelisojeah commented on webpack.prod.js:
```diff
@@ -0,0 +1,66 @@
+const path = require('path');
+const merge = require('webpack-merge');
+const webpack = require('webpack');
+const ExtractTextPlugin = require('extract-text-webpack-plugin');
```
This is deprecated

GerrardE replied:
Feedback has been implemented

fidelisojeah commented on webpack.prod.js:
```diff
@@ -0,0 +1,66 @@
+const HtmlWebpackPlugin = require('html-webpack-plugin');
+const baseConfig = require('./webpack.base.js');
+
+module.exports = merge(baseConfig, {
```
I'm not convinced with your webpack set up.

A whole lot of things you are doing have been changed with better set up.

webpack 4 has an `optimization` object which you should modify and work with.


also the `bail` object to terminate early on error

GerrardE replied:
Feedback Implemented, thanks

fidelisojeah commented on webpack.dev.js:
```diff
@@ -0,0 +1,57 @@
+      title: 'Authors Haven',
+      template: './client/index.html'
+    }),
+  ],
```
include the `HotModuleReplacementPlugin`

GerrardE replied:
Feedback has been implemented

fidelisojeah reviewed (changes requested):
need some changes here

also,

no need for the `client` folder

put all your work in a `src` folder which would be located in the root location

GerrardE committed:
ch-[166272288]: Set up react project
>
>
> - Refactor webpack configurations for production and development using DRY principle

[#166272288]

GerrardE committed:
ch-[166272288]: Set up react project

- Refactor eslintrc

[#166272288]

fidelisojeah commented on package.json:
```diff
@@ -0,0 +1,90 @@
+    "@babel/plugin-proposal-class-properties": "^7.4.4",
+    "@babel/plugin-proposal-decorators": "^7.0.0",
+    "@babel/plugin-proposal-do-expressions": "^7.0.0",
+    "@babel/plugin-proposal-export-default-from": "^7.0.0",
```
You are not using any of these

Please get rid of them immidiately.

fidelisojeah commented on webpack.base.js:
```diff
@@ -0,0 +1,80 @@
+      }
+    ]
+  },
+  optimization: {
```
optimization should only be in the production build. Move this.

use the `TerserPlugin` and `OptimizeCSSAssetsPlugin` in the `optimization. minimizer `.

GerrardE replied:
Feedback Implemented

fidelisojeah commented on app.js:
```diff
@@ -0,0 +1,19 @@
+import express from 'express';
```
Please get rid of this ASAP

GerrardE replied:
Feedback implemented

fidelisojeah commented on package.json:
```diff
@@ -0,0 +1,90 @@
+  },
+  "homepage": "https://github.com/andela/kingsmen-ah-frontend#readme",
+  "devDependencies": {
+    "@babel/core": "^7.4.5",
```
this and a whole lot of others should be in `dependencies`

fidelisojeah commented on webpack.base.js:
```diff
@@ -0,0 +1,80 @@
+        enforce: 'pre',
+        test: /\.jsx|js?$/,
+        exclude: /node_modules/,
+        loader: 'eslint-loader'
```
```suggestion
        use:[{
          options:{
            eslintPath: require.resolve('eslint'),
          },
          loader: 'eslint-loader'
        }]
        
```

fidelisojeah reviewed (changes requested):
Please correct this

GerrardE committed:
ch-[166272288]: Set up react project

- Refactor packageJson and webpackConfig

[#166272288]

despeauxz commented on webpack.base.js:
```diff
@@ -0,0 +1,62 @@
+          { loader: 'css-loader', options: { sourceMap: true } },
+          { loader: 'less-loader', options: { sourceMap: true } },
+          { loader: 'node-sass', options: { sourceMap: true } }
+        ]
```
Nice job, might be wrong but I think `node-sass` is just dependent on the other loaders like `css-loader`, `style-loader`, `sass-loader`. That should be replaced with the `sass-loader`

fidelisojeah replied:
Actually, I think you are correct here. @despeauxz 

In addition, we wouldn't be using `less` so that loader should be gotten rid of.

GerrardE replied:
Thanks for the feedback. I have implemented it.

fidelisojeah replied:
Since you are already doing this, I have assigned #166272293 to you

GerrardE committed:
ch-[166272288]: Set up react project

- Refactor packageJson and webpackConfig

[#166272288]

despeauxz reviewed (commented):
Great work so far, but as our initial discussion we aren't able to render the dist `index.html` file. Maybe creating a server to start the `dist` folder would be nice

fidelisojeah commented on package.json:
```diff
@@ -0,0 +1,76 @@
+    "uglifyjs-webpack-plugin": "^2.0.1",
+    "webpack": "^4.35.2",
+    "webpack-cli": "^3.3.5",
+    "webpack-dev-server": "^3.7.2",
```
`devDependency`

fidelisojeah commented on package.json:
```diff
@@ -0,0 +1,76 @@
+    "babel-loader": "^8.0.6",
+    "babel-plugin-lodash": "^3.3.2",
+    "babel-plugin-module-resolver": "^3.0.0",
+    "babel-plugin-react-css-modules": "^3.3.2",
```
Are you using this?

fidelisojeah commented on package.json:
```diff
@@ -0,0 +1,76 @@
+    "css-loader": "^0.28.7",
+    "debug": "^4.1.1",
+    "dotenv": "^8.0.0",
+    "express": "^4.17.1",
```
We aren't using this

fidelisojeah commented on package.json:
```diff
@@ -0,0 +1,76 @@
+    "babel-plugin-react-css-modules": "^3.3.2",
+    "compression-webpack-plugin": "^3.0.0",
+    "css-loader": "^0.28.7",
+    "debug": "^4.1.1",
```
or this

fidelisojeah commented on webpack.base.js:
```diff
@@ -0,0 +1,62 @@
+    rules: [
+      {
+        enforce: 'pre',
+        test: /\.jsx|js?$/,
```
```suggestion
        test: /\.(js|jsx)$/,
```

GerrardE replied:
Feedback has been implemented. Thanks

GerrardE committed:
ch-[166272288]: Set up react project

- Refactor packageJson and webpackConfig

[#166272288]

despeauxz reviewed (approved):
Nice job
